### PR TITLE
Banc Sabadell

### PR DESCRIPTION
**Pull request by mikz**
This is working Banc Sabadell integration.

* integrates rspec and specs out the sabadell integration
* adds abstract http client, that should be able to use different drivers on different platforms
* sabadell integration, that does not inherit from bank
* using [fake_bsmobil](https://github.com/mikz/fake_bsmobil) to implement fake api for testing
* travis ci

I'm not sure if it is ready, there is plenty things I'd like to do. Add more docs, document it a bit, etc.
I noticed few tiny things like year in transactions is wrong, but have no power to fix that now.

The code requires Ruby 2.2. Which is probably not cool for rbx/jruby/rubymotion support, what MRI version are you targeting?

### CODE REVIEW

File: lib/bank_scrap/banks/banc_sabadell.rb [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+      @http_client.headers.merge!(
+          'Accept' => 'application/vnd.idk.bsmobil-v2+json',
+          'Content-Type' => 'application/json'
+      )
+
+      @session = Session.new(self)
+
+      @nie = nie or raise InvalidCredentialsError, "NIE can't be #{nie.inspect}"
+      @pin = pin or raise InvalidCredentialsError, "NIE can't be #{nie.inspect}"
```
Comment (jguitar) [inline lib/bank_scrap/banks/banc_sabadell.rb]: You made a copy/paste error here, it should be _PIN can't be..._

**Comment by javiercr** [comment]
Hello @mikz,

First of all, sorry for not reviewing this earlier and thank you for your awesome contribution. It's clear you put a lot of time into this.

I've reviewed the PR and I would like to suggest to split this in different PRs after we discuss each change. There are 53 files modified and the whole structure of the gem has been modified, so I think we would do better if we discuss each big change individually. Here are my thoughts regarding each of them:

---
### Splitting into multiple gems (under the same repo) 

This is probably the biggest change you made. I'm a bit confused since is the first time I see an approach like this.  

The main problem with the single gem approach is the dependencies. For instance ING requires `rmagick`, which often is painful to install, so if you just want to use BBVA you still need to install the rmagick dependency. On your PR I've seen that the new `bank_scrap-core` has all the other gems as dependencies, so I guess this doesn't solve this problem, does it?

Could you elaborate on what are the benefits of this multi-gem approach?

### Abstract HTTP client
I really like this one. From my experience with BBVA I had to try 4 different clients until I found one (`mechanize`) that allowed me to replicate the exact network traffic that the API was expecting. Other banks could have similar issues and maybe `mechanize` is not the best client for them. Therefore I think this is a great idea.

However as the main gem has the bank's gem as dependencies, and those banks as the mechanize gem as dependency, right now if I install the main `bank_scrap` gem I will be forced to install the `bank-_scrap-mechanize` and `mechanize` gems. How are you planning to make this really optional?

### Rspec and testing
This was something we were really needing, so kudos for it. I still need to checkout if I can replicate what you did with [fake_bsmobil](https://github.com/mikz/fake_bsmobil) with BBVA.

---
Finally as this is a big change to the project I think we should try to come up with the biggest consensus possible between all the contributors. I'm gonna ping those who appear as contributors in the repo but any opinion is welcome: @ismaGNU @raulmarcosl @ferblape @javierarce @hopsor.

Thank you!

**Comment by mikz** [comment]
Thanks very much for the feedback. The intention was to spark the discussion about substantial changes that will be required to make it multi platform.
I'm on vacation now so I'll reply in a week or so. Cheers.

**Comment by ferblape** [comment]
@javiercr I agree with you @mikz did a great job. Thanks for the effort. I'd go also for separated gems, makes more sense to keep dependencies and code simple and issues can be split for each bank.

**Comment by mikz** [comment]
@javiercr thanks very much for taking the time. 

The commits in the PR make sense on its own. The biggest one is just a kitchen sink with all the integrations split. It is there just to show how it would look like. The Banc Sabadell integration works without it. It definitely makes sense to tackle each of the issues separately so I'll create different PR for the split later and we can continue discussion there.

-----------

## Splitting into multiple gems (under the same repo)

Agree that the dependencies are the issue. That is the main reason.
I want to use bank_scrap in RubyMotion and there is no RMagick or other gems like mechanize. 

### BankScrap::Core
`bank_scrap-core` has no dependencies, see [`bank_scrap-core.gemspec`](https://github.com/mikz/bank_scrap/blob/215212663f944547e44cd68d26db14112d78e50b/bank_scrap-core/bank_scrap-core.gemspec). Integrations depend on `bank_scrap-core` as it implements some basic classes. The [README](https://github.com/mikz/bank_scrap/blob/215212663f944547e44cd68d26db14112d78e50b/bank_scrap-core/README.md) says:
> Bank Scrap Core is zero dependency core that providers models and basic structures for building bank integrations.

### Dependencies
I imagine BankScrap aims to be a library. Not a CLI utility. Therefore it should aim to be as much dependency abstract as it can be and if possible do not have some dependencies at all.
The core aims to be totally dependency free so it can be installed on any platform.

Every integration declares just own dependencies. See [`bank_scrap-bankinter.gemspec`](https://github.com/mikz/bank_scrap/blob/215212663f944547e44cd68d26db14112d78e50b/bank_scrap-bankinter/bank_scrap-bankinter.gemspec) for example. It depends on `execjs` and `nokogiri`. So it does solve problem of RMagick.

That might lead to some conflicts between integrations (like one locking too tight to a version of a dependency), but that can be solved by loading them all in tests (so they won't load because of bundler), and defining some constants like `NOKOGIRI_DEPENDENCY` in shared `Gem::Specification` called [`BankScrap::Specification`](https://github.com/mikz/bank_scrap/blob/215212663f944547e44cd68d26db14112d78e50b/bank_scrap-base.rb).
Which is there to have common base for all the integrations, like license, url, ruby version, rspec, etc.
It is not the best name, but it is just a prototype.

### Plugins / Integrations
I changed the [`#find_bank_class_for(bank_name)`](https://github.com/mikz/bank_scrap/blob/215212663f944547e44cd68d26db14112d78e50b/bank_scrap-cli/lib/bank_scrap/cli.rb#L79-L84) to not only rely on autoload, but require the dependency first.

This can be further optimised by using [`Gem::find_files(glob)`](http://docs.ruby-lang.org/en/2.0.0/Gem.html#method-c-find_files) like [Rubygems plugins](http://docs.ruby-lang.org/en/2.0.0/Gem.html#module-Gem-label-RubyGems+Plugins)
> As of RubyGems 1.3.2, RubyGems will load plugins installed in gems or $LOAD_PATH. Plugins must be named ‘rubygems_plugin’ (.rb, .so, etc) and placed at the root of your gem’s #require_path. Plugins are discovered via Gem::find_files then loaded. Take care when implementing a plugin as your plugin file may be loaded multiple times if multiple versions of your gem are installed.

So the CLI can for example print all the available integrations.

### Future

In the future I'd imagine that there could be several abstract XML parsers, JS executors, HTTP clients and it would be on the platform to decide which to choose. For example good luck using Nokogiri on Rubinius. 
Also, with the gem approach there can be for example 'base' ING integration and then other two that extend it with some platform specific way.


## Abstract HTTP client

I needed the Abstract HTTP Client for RubyMotion. And I really did not need whole `mechanize`. Actually it was more work to use `mechanize` than to use plain Net::HTTP.

Some integrations were using mechanize, so I kept that by just wrapping it in `bank_scrap-mechanize` and declaring that as a dependency in each integration. It does not follow the API of abstract http client. Not sure if it can, but that is really not needed now.

What should installing `bank_scrap` do is question that I can't answer. Should it be the CLI, should it include some integrations, should it be just the core?

Right now, I was trying to keep the layout where installing `bank_scrap` gives you this bag of tools and you can use it all, but you can also cherry pick just individual integrations as libraries.

I'd love to make mechanize optional, but that would require changing the integrations, which I did not want to do without test coverage. But when it would be optional, why would you need it at all? It uses `Net::HTTP` underneath.  As long as integrations will depend on mechanize, it can't be made optional.

But you can just install integrations that do not require mechanize and use them without mechanize (that is my plan actually how to use it on RubyMotion).

## Rspec and testing

Using the fake server was awesome experience. Writing the server took some time, but then the integration took just a bit.

We could extract some foundation from `fake_bsmobil` and use it for other integrations.

-------


This PR was meant to include Banc Sabadell integration and show splitting of the gems to spark some discussion. I'm glad that it finally happened. 

I'll take out the gem splitting and just include the Banc Sabadell so it can be merged soon and issue another PR for the split. There we can discuss how to do it (one by one, all at once, what makes sense, ...).

Wdyt?

**Comment by mikz** [comment]
@javiercr rebased this and moved the split to #30 

This now contains just the Banc Sabadell integration.

**Pull request closed without merging**